library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java

```diff
@@ -0,0 +1,167 @@
+
+@Mixin(ItemRenderer.class)
+public abstract class ItemRendererMixin {
+	@Unique private final ThreadLocal<MatrixStack> quilt$matrices = ThreadLocal.withInitial(MatrixStack::new);
+
+	@Shadow public float zOffset;
+
+	@Inject(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At("HEAD"), cancellable = true)
+	private void quilt$customizeItemOverlay(TextRenderer renderer, ItemStack stack, int x, int y, String countLabel,
+											CallbackInfo ci) {
+		if (stack.isEmpty()) {
+			return;
+		}
+
+		var item = stack.getItem();
+
+		var matrices = quilt$matrices.get();
+		matrices.push();
+		matrices.translate(x, y, 0);
+
+		if (item.preRenderOverlay(matrices, renderer, this.zOffset, stack)) {
+			if (isItemOverlayRenderingCustomized(item)) {
+				ci.cancel();
+				renderCustomGuiItemOverlay(matrices, renderer, stack, countLabel);
+			}
+		} else {
+			ci.cancel();
+		}
+
+		matrices.pop();
+	}
+
+	@Redirect(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At(value = "NEW", target = "net/minecraft/client/util/math/MatrixStack"))
+	private MatrixStack quilt$avoidMatrixStackAllocation() {
+		return quilt$matrices.get(); // no need to push now, matrices are only modified when count label is rendered
+	}
+
+	@Redirect(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At(value = "INVOKE", target = "Lnet/minecraft/client/util/math/MatrixStack;translate(DDD)V"))
+	private void quilt$renderCountLabel_pushMatrixStack(MatrixStack instance, double x, double y, double z) {
+		instance.push();
+		instance.translate(x, y, z);
+	}
+
+	@Inject(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At(value = "INVOKE", target = "Lnet/minecraft/client/render/VertexConsumerProvider$Immediate;draw()V",
+					shift = At.Shift.AFTER), locals = LocalCapture.CAPTURE_FAILHARD)
+	private void quilt$renderCountLabel_popMatrixStack(TextRenderer renderer, ItemStack stack, int x, int y, String countLabel,
+													   CallbackInfo ci, MatrixStack matrices) {
+		matrices.pop();
+	}
+
+	@Inject(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At("TAIL"), locals = LocalCapture.CAPTURE_FAILHARD)
+	private void quilt$invokePostRenderOverlay(TextRenderer renderer, ItemStack stack, int x, int y, String countLabel,
+											   CallbackInfo ci, MatrixStack matrices) {
+		if (stack.isEmpty()) {
+			return;
+		}
+
+		matrices.push();
+		matrices.translate(x, y, 0);
+		stack.getItem().postRenderOverlay(matrices, renderer, this.zOffset, stack);
+		matrices.pop();
+	}
+
+	@Unique
+	private boolean isItemOverlayRenderingCustomized(Item item) {
```

Comment:
Does adding `@Unique` make sure that the method never collides? 
Does it salt the method or something? If it doesn't I suppose you could prefix it with quilt$ to ensure nothing weird happens.

Reply:
![image](https://user-images.githubusercontent.com/46004440/180822157-fff6a287-22ef-4aab-91b2-a699a20cfcda.png)
Yeap javadocs confirms that to be the case.